onnxruntime/core/providers/cuda/nn/pool.cc

```diff
@@ -152,20 +153,37 @@ Status Pool<T, PoolType, NHWC>::ComputeInternal(OpKernelContext* context) const
   const TensorShape& x_shape = X->Shape();
   const auto x_dims = x_shape.GetDims();
 
-  if (x_shape.NumDimensions() < 3) {
-    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Input dimension cannot be less than 3.");
-  }
-
   auto kernel_shape = pool_attrs_.kernel_shape;
   auto pads = pool_attrs_.pads;
   auto strides = pool_attrs_.strides;
+  if (x_shape.NumDimensions() < 3) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Input dimension cannot be less than 3.");
+  }
+  if constexpr (NHWC) {
+    if (pool_attrs_.global_pooling) {
+      // Set kernel size to the spatial dimension of input tensor.
+      // The first dim of x_dims is batch size(N).
+      // The last dim of x_dims is channel(C).
+      // Put the other part in kernel_shape
+      kernel_shape.assign(x_dims.begin() + 1, x_dims.end() - 1);
+      pads.assign(kernel_shape.size(), 0);
```

Comment:
Then how could the old code work?

Reply:
```c++
 if (pool_attrs_.global_pooling) {
    kernel_shape.assign(x_dims.begin() + 2, x_dims.end());
    pads.assign(kernel_shape.size(), 0);
    strides.assign(kernel_shape.size(), 1);
  }
```

Reply:
We have test case for that? How did it pass?

Reply:
AFAIK there are no "pads" for Global pooling - https://onnx.ai/onnx/operators/onnx__GlobalMaxPool.html.

Reply:
If pads is 2x the size of kernel_shape the function `CudnnPoolingDescriptor::Set` is broken as well. It's passing only kernel_shape.size() elements of pads to `SetPoolingNdDescriptorHelper`.

Reply:
This change passes all tests for me added locally added to PR #19889 

```  if (kernel_shape.size() < 2) {
    // cuDNN only takes 4D or 5D input, so pad dimensions if needed
    if (Layout == LAYOUT_NHWC) {
      x_dims_cudnn.insert(x_dims_cudnn.begin() + 1, 1);
      y_dims_cudnn.insert(y_dims_cudnn.begin() + 1, 1);
      pads.insert(pads.begin(), 0);
      kernel_shape.insert(kernel_shape.begin(), 1);
      strides.insert(strides.begin(), 1);
    } else {
      x_dims_cudnn.insert(x_dims_cudnn.begin() + 2, 1);
      y_dims_cudnn.insert(y_dims_cudnn.begin() + 2, 1);
      pads.insert(pads.begin(), 0);
      kernel_shape.insert(kernel_shape.begin(), 1);
      strides.insert(strides.begin(), 1);
    }
  }
```

Reply:
#19889 supersedes this PR by fixing a few more NHWC related issues and enabling all nhwc for all pooling operations supported by the EP.

Reply:
For the else part:

      x_dims_cudnn.insert(x_dims_cudnn.begin() + 2, 1);
      y_dims_cudnn.insert(y_dims_cudnn.begin() + 2, 1);
      pads.insert(pads.begin(), 0);
      kernel_shape.insert(kernel_shape.begin(), 1);
      strides.insert(strides.begin(), 1);

This is a deviation from what we previously did. This adds the "padding" feature dim to the "beginning" (i.e.) makes NCH -> NCWH. Previously, we were doing NCW -> NCWH. Any thoughts on this ? Is one approach better than the other from CuDNN point of view ?

Reply:
In Conv, default is padded to NCD1 instead of NC1D (user need `cudnn_conv1d_pad_to_nc1d` option to enable the later:
https://onnxruntime.ai/docs/execution-providers/CUDA-ExecutionProvider.html#convolution-input-padding
I think it is better to follow similar logic.

Reply:
I agree.

Reply:
I updated my code. But I think it is still not right. I think CuDNN only support symmetric padding, which means the padding size at the beginning and end should be the same. Therefore here the length of "pad" should be equal to the length of kernel_shape.

Reply:
@snnn, that's discussed in https://github.com/microsoft/onnxruntime/pull/19889#discussion_r1525183243. I think that PR is almost ready (may be tomorrow), that PR can address this bug as well.

Reply:
Maybe we can just fix the NCHW regression (for the patch) and leave anything related to NHWC un-implemented that can be filled in by the other PR. In any case, I don't think we have a ton of NHWC users just yet that will be broken by the patch release.

Reply:
I think either is fine. If we gave up this one,  we should cherry-pick #19889 to the patch release. That's PR is not very big. So I am not worried.

Reply:
After the bugs are fixed ,we should refactor the SetPoolingNdDescriptorHelper function a little: change the three array type parameters to std::span, and check if all their lengths equal. I can take the work after #19889 is merged.

Reply:
And in CUDAExecutionProvider::GetCapability function https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/providers/cuda/cuda_execution_provider.cc we should check if paddings are symmetric when the op is a pooling op but it is not global pooling.